kernel: spinlock: expose LOCKED pattern in public API

fgrandel opened the pull request:
This PR proposes to expose the LOCKED(&mylock) { ...do something with lock held... } pattern as a public API.

While the LOCKED pattern is universally useful and produces nice and readable code it can easily provoke deadlock if abused. 

This is a controversial thing to do so the PR should be understood as an RFC the purpose of which is to either provide this pattern in userspace or document in a discoverable and linkable way why it should not be public.

Commit:
kernel: spinlock: expose LOCKED macro as public API

While the LOCKED pattern is universally useful it can be misused. This
change therefore exposes the LOCKED pattern with extensive usage
documentation to reduce the risk of abuse or unintended deadlock.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

peter-mitsis commented on include/zephyr/spinlock.h:
```diff
@@ -29,6 +29,35 @@ extern "C" {
+ * This basically expands to:
+ *
+ * @code{.c}
+ * for (k_spinlock_key_t key = k_spin_lock(&mylock); k_spin_unlock(&mylock, key); ...) {
```
Minor detail: The expansion is incorrect. The compare element in a for-loop is executed before we actually enter the body of the for-loop. As a result, there is a discrepancy with the claim above on line 40.

fgrandel replied:
@peter-mitsis Should be correct now. Could you have another look?

peter-mitsis commented on include/zephyr/spinlock.h:
```diff
@@ -29,6 +29,35 @@ extern "C" {
+ * section with a single point of exit in the end. As it uses a for-loop
+ * internally, you MUST NOT EXIT ANYWHERE FROM THE BLOCK (e.g. using a break or
+ * return statement) otherwise the spinlock will not be released.
+ */
```
There is an additional caveat that might be worth noting.  As this for-loop has only a single iteration, using a _continue_ statement may allow one to safely leave the block. This is on the surface somewhat counter intuitive, so it may be worthwhile to provide a macro that expands to _continue_ that can be used to safely leave the block.

fgrandel replied:
Good point. See whether you like my proposal.

peter-mitsis commented on include/zephyr/spinlock.h:
```diff
@@ -29,6 +29,35 @@ extern "C" {
+ * internally, you MUST NOT EXIT ANYWHERE FROM THE BLOCK (e.g. using a break or
+ * return statement) otherwise the spinlock will not be released.
+ */
+#define WITH_SPINLOCK(lck)                                                                         \
```
Possible name change? K_HOLD_SPINLOCK ?

dcpleung replied:
`SPINLOCK_GUARDED`? Since you are creating a new macro, it is a good idea to make it more descriptive.

fgrandel replied:
@peter-mitsis @dcpleung K_SPINLOCK_GUARDED maybe? I think the K_ prefix makes sense here.

dcpleung reviewed (commented):
Also, one needs to be very careful on where the spinlock is located. By default, any `static` global variables reside in `.data` or `.bss` section which is not accessible from userspace.  One needs to explicitly mark them with `K_APP_BMEM()` or `K_APP_DMEM()`, and also the necessary bits with partitions and memory domains.

fgrandel commented:
> Also, one needs to be very careful on where the spinlock is located. By default, any `static` global variables reside in `.data` or `.bss` section which is not accessible from userspace.

@dcpleung Not sure how this plays a role here? We're not defining any spinlock here nor do we introduce any new usage pattern. The variables we introduce are on the stack and should be accessible, right? We just introduce some syntactic sugar on top of code which the developer would have written anyway.

In other words: If a spinlock is placed into inaccessible memory then it would fail with or without this definition. I'd think it's not the responsibility of a helper like this to ensure correct section and memory domain placement of the spinlock. Do you agree?

dcpleung commented:
> > Also, one needs to be very careful on where the spinlock is located. By default, any `static` global variables reside in `.data` or `.bss` section which is not accessible from userspace.
> 
> @dcpleung Not sure how this plays a role here? We're not defining a spinlock. The additional loop variables should be on the stack and accessible. It's just some syntactic sugar.
> 
> In other words: If a spinlock is placed into inaccessible memory then it would fail with or without this definition. Do you agree?

I was thinking maybe we could add a comment about variable placement for userspace in the macro description.

andyross commented on include/zephyr/spinlock.h:
```diff
@@ -29,6 +29,56 @@ extern "C" {
+ *
+ * See @ref K_SPINLOCK_GUARDED_SECTION for details.
+ */
+#define K_SPINLOCK_LEAVE_GUARDED_SECTION continue
```
Maybe just "K_SPINLOCK_BREAK", since that's what it's doing?  Not a big fan of long jargon terms like "GUARDED_SECTION" personally.

andyross commented on include/zephyr/spinlock.h:
```diff
@@ -29,6 +29,56 @@ extern "C" {
+ *
+ * @note The code block must execute to its end or be left by calling
+ * K_SPINLOCK_LEAVE_GUARDED_SECTION. Otherwise, e.g. if exiting the block with
+ * a break or return statement, the spinlock will not be released on exit.
```
Might want to add `goto` to the list, as "linux goto style" error handling is fairly common in Zephyr and is incompatible with this trick too.

andyross reviewed (dismissed):
This is a trick that IMHO seems a lot more pleasing within a single C file than it does as an API.  But I liked it where it started and I'll continue to like it as an API, I'm just not willing to commit to big fights over things.  +1 from me in whatever form it takes.

One note is that there actually is a GCC extension that allows for cleanup/destructors on exit from a block a-la C++.  We could use that to safely get break/goto exit handling to work, but it would actually be more useful to detect and panic on such a case, since most users will have access to it and it guarantees we don't end up with non-portable code.

Commit:
kernel: spinlock: remove internal function from docs

Hides a function marked internal from the documentation.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
kernel: spinlock: add assertion to protect K_SPINLOCK

Adds an assertion for the GCC toolchain to check whether the K_SPINLOCK
block was left without unlocking the spinlock, e.g. by calling break,
return or goto.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

fgrandel commented:
> I was thinking maybe we could add a comment about variable placement for userspace in the macro description.

@dcpleung ok, makes sense. I'm not an MPU expert but tried to add some wording. Maybe you can review again and check whether that fits...

> This is a trick that IMHO seems a lot more pleasing within a single C file than it does as an API. [...] I'm just not willing to commit to big fights over things.

@andyross If the end result doesn't feel 100% right, then let me know and we'll just scrap it, no matter how much time has been invested on my side. I mostly see this as a learning opportunity anyway.

> [...] there actually is a GCC extension that allows for cleanup/destructors on exit [...] it would actually be [...] useful to detect and panic on such a case, since most users will have access to it and it guarantees we don't end up with non-portable code.

@andyross I like this proposal as it makes the pattern much safer. I tried to implement it in a separate commit for review. But please bare with me if I got it wrong - this is all rather unfamiliar territory for me (albeit very interesting). I tested this on real hardware and AFAICS it works as intended.

> Maybe just "K_SPINLOCK_BREAK"

@andyross Fixed - also renamed the actual pattern to K_SPINLOCK which seams much cleaner and still says it all. I like this even better now.

> Might want to add goto to the list

@andyross Done.

andyross commented on kernel/include/kernel_internal.h:
```diff
@@ -227,10 +227,8 @@ void z_mem_manage_init(void);
-			!__i.key;					\
-			k_spin_unlock(lck, __key), __i.key = 1)
+/* Alias of K_SPINLOCK to keep kernel code concise. */
+#define LOCKED K_SPINLOCK
```
No reason the kernel can't use the same symbol everyone else does, FWIW.

fgrandel replied:
Fixed.

andyross reviewed (dismissed):
Refresh +1.  One nitpick.

Commit:
kernel: spinlock: LOCKED -> K_SPINLOCK

Let the kernel use the new K_SPINLOCK macro and remove the alias.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

fgrandel commented:
@dcpleung @andyross @peter-mitsis Should be ready for final review/merge now.

andyross reviewed (approved):
Still +1 from.  I was genuinely expecting more argument.  Nice.

fgrandel commented:
Thanks everyone for your support!
